File: android/app/src/main/kotlin/com/hr/app/data/sync/OutboxHttpSender.kt
package com.hr.app.data.sync

import android.util.Log
import com.hr.app.BuildConfig
import com.hr.app.data.local.OutboxEntry
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Sends outbox entries over HTTP.
 *
 * Deliberately generic rather than typed: an outbox entry already carries a method, a path and a
 * serialised body. Routing each entry through its typed generated-client method would mean a
 * `when` over every mutation in the product, which is a maintenance burden that grows with every
 * feature and buys nothing — the payload was serialised from a typed object when it was enqueued.
 *
 * The status-code mapping is the important part, and follows docs/sync-protocol.md §4.4.
 */
@Singleton
class OutboxHttpSender
    @Inject
    constructor(
        private val client: OkHttpClient,
        private val json: Json,
    ) : OutboxSender {
        override suspend fun send(entry: OutboxEntry): SendOutcome =
            withContext(Dispatchers.IO) {
                try {
                    val request =
                        Request.Builder()
                            .url(BuildConfig.API_BASE_URL.trimEnd('/') + "/" + entry.path.trimStart('/'))
                            .method(entry.httpMethod, entry.payload.toRequestBody(JSON_MEDIA_TYPE))
                            // The contract that makes unlimited retry safe. Generated once when
                            // the entry was created, never regenerated here.
                            .header("Idempotency-Key", entry.idempotencyKey)
                            .build()

                    client.newCall(request).execute().use { response ->
                        val body = response.body?.string().orEmpty()
                        when {
                            response.isSuccessful -> SendOutcome.Confirmed

                            // Someone else acted first. That is a settled outcome, not a failure —
                            // retrying would never change it, and surfacing it as an error would
                            // be misleading.
                            response.code == 409 && errorCode(body) == "ALREADY_DECIDED" ->
                                SendOutcome.Confirmed

                            response.code == 401 -> SendOutcome.AuthenticationRequired

                            response.code == 429 || response.code >= 500 -> SendOutcome.Retryable

                            response.code in 400..499 ->
                                SendOutcome.Rejected(errorCode(body), errorMessage(body))

                            else -> SendOutcome.Retryable
                        }
                    }
                } catch (e: IOException) {
                    // Network unavailable, DNS failure, timeout. Always retryable.
                    Log.d(TAG, "Network failure sending outbox entry ${entry.id}: ${e.message}")
                    SendOutcome.Retryable
                } catch (e: Exception) {
                    // A malformed entry would fail identically forever, so retrying is pointless.
                    Log.w(TAG, "Unrecoverable failure sending outbox entry ${entry.id}", e)
                    SendOutcome.Rejected("CLIENT_ERROR", e.message)
                }
            }

        /** Extracts `error.code` from the standard envelope (docs/03-architecture.md §9). */
        private fun errorCode(body: String): String? = errorField(body, "code")

        private fun errorMessage(body: String): String? = errorField(body, "message")

        private fun errorField(
            body: String,
            field: String,
        ): String? =
            runCatching {
                json.parseToJsonElement(body).jsonObject["error"]?.jsonObject?.get(field)?.jsonPrimitive?.content
            }.getOrNull()

        private companion object {
            const val TAG = "OutboxHttpSender"
            val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
        }
    }

File: ios/Sources/HRCore/Sync/OutboxHTTPSender.swift
import Foundation
import OSLog

/// Sends outbox entries over HTTP.
///
/// The status mapping is the whole substance of this type, and it follows
/// docs/sync-protocol.md §4.4. It is kept deliberately identical to `OutboxHttpSender` on Android:
/// the two platforms implement one written protocol, and a divergence here would show up as
/// "leave requests silently vanish on iPhones" long after the change that caused it.
public struct OutboxHTTPSender: OutboxSender {
    private let client: HTTPClient
    private let log = Logger(subsystem: "io.hrapp", category: "OutboxSender")

    public init(client: HTTPClient) {
        self.client = client
    }

    public func send(_ entry: OutboxEntry) async -> SendOutcome {
        let outcome = await client.request(
            method: entry.httpMethod,
            path: entry.path,
            body: Data(entry.payload.utf8),
            // `idempotencyKey`, not `id`. They are distinct fields on an entry, and using the row
            // id here would still be *a* stable key — but a different one from Android's, so the
            // same logical retry would look like a new request to the server. The contract only
            // works if both clients send the field the protocol names.
            //
            // Generated once when the entry was enqueued and never regenerated here: regenerating
            // on retry is precisely what turns a lost response into a duplicate record.
            idempotencyKey: entry.idempotencyKey
        )

        switch outcome {
        case .failure(.unauthenticated):
            return .authenticationRequired

        case .failure(let error) where error.isRetryable:
            log.debug("Retryable failure sending \(entry.id): \(String(describing: error))")
            return .retryable

        case .failure(let error):
            // A malformed entry would fail identically forever, so retrying is pointless.
            log.warning("Unrecoverable failure sending \(entry.id): \(String(describing: error))")
            return .rejected(code: "CLIENT_ERROR", message: String(describing: error))

        case .success(let response):
            return classify(response, entry: entry)
        }
    }

    /// Internal rather than private so the mapping can be tested directly.
    ///
    /// It is the substance of this type and every branch matters, but reaching all of them through
    /// a real ``HTTPClient`` would mean stubbing `URLSession` to return eight different status
    /// codes — machinery that tests the stub more than the mapping.
    func classify(_ response: HTTPClient.Response, entry: OutboxEntry) -> SendOutcome {
        if response.isSuccess { return .confirmed }

        let envelope = APIErrorEnvelope.decode(from: response.body)

        switch response.status {
        // Someone else acted first. That is a settled outcome, not a failure — retrying would
        // never change it, and surfacing it as an error would be misleading to the user.
        case 409 where envelope?.code == "ALREADY_DECIDED":
            return .confirmed

        case 401:
            return .authenticationRequired

        case 429, 500...:
            return .retryable

        case 400...499:
            return .rejected(code: envelope?.code, message: envelope?.message)

        default:
            return .retryable
        }
    }
}
